File: src/coords.js
const directions = [
  { x: 1, y: -1, z: 0 },
  { x: 1, y: 0, z: -1 },
  { x: 0, y: 1, z: -1 },
  { x: -1, y: 1, z: 0 },
  { x: -1, y: 0, z: 1 },
  { x: 0, y: -1, z: 1 },
];

const key = ({ x, y, z }) => `${x},${y},${z}`;

const parseKey = (k) => {
  const [x, y, z] = k.split(',').map(Number);
  return { x, y, z };
};

const add = (a, b) => ({ x: a.x + b.x, y: a.y + b.y, z: a.z + b.z });

const equals = (a, b) => a.x === b.x && a.y === b.y && a.z === b.z;

const neighbors = (coord) => directions.map((d) => add(coord, d));

const directionIndex = (from, to) => {
  const diff = { x: to.x - from.x, y: to.y - from.y, z: to.z - from.z };
  return directions.findIndex((d) => equals(d, diff));
};

const distance = (a, b) =>
  (Math.abs(a.x - b.x) + Math.abs(a.y - b.y) + Math.abs(a.z - b.z)) / 2;

module.exports = {
  directions,
  key,
  parseKey,
  add,
  equals,
  neighbors,
  directionIndex,
  distance,
};

File: src/coords.py
from dataclasses import dataclass
from typing import List, Tuple


@dataclass(frozen=True)
class Coord:
    x: int
    y: int
    z: int

    def key(self) -> str:
        return f"{self.x},{self.y},{self.z}"


DIRECTIONS: List[Coord] = [
    Coord(1, -1, 0),
    Coord(1, 0, -1),
    Coord(0, 1, -1),
    Coord(-1, 1, 0),
    Coord(-1, 0, 1),
    Coord(0, -1, 1),
]


def parse_key(k: str) -> Coord:
    x, y, z = map(int, k.split(","))
    return Coord(x, y, z)


def add(a: Coord, b: Coord) -> Coord:
    return Coord(a.x + b.x, a.y + b.y, a.z + b.z)


def neighbors(coord: Coord) -> List[Coord]:
    return [add(coord, d) for d in DIRECTIONS]


def key(coord: Coord) -> str:
    return coord.key()


def direction_index(from_coord: Coord, to_coord: Coord) -> int:
    diff = Coord(to_coord.x - from_coord.x, to_coord.y - from_coord.y, to_coord.z - from_coord.z)
    try:
        return DIRECTIONS.index(diff)
    except ValueError:
        return -1


def distance(a: Coord, b: Coord) -> int:
    return (abs(a.x - b.x) + abs(a.y - b.y) + abs(a.z - b.z)) // 2
